core/builtin.py

```diff
@@ -604,28 +604,51 @@ def Cd(argv, mem, dir_stack):
   return 0
 
 
-WITH_PREFIX = 1
-WITHOUT_PREFIX = 2
+WITH_LINE_NUMBERS = 1
+WITHOUT_LINE_NUMBERS = 2
 SINGLE_LINE = 3
+TILDE_PREFIX_FORM = 1
+LONG_FORM = 2
 
-def _PrintDirStack(dir_stack, style):
+def _PrintDirStack(dir_stack, mem, style, form):
   """Helper for 'dirs'."""
 
-  if style == WITH_PREFIX:
+  if style == WITH_LINE_NUMBERS:
     for i, entry in enumerate(dir_stack.Iter()):
-      print('%2d  %s' % (i, entry))
+      if form == TILDE_PREFIX_FORM:
```

Comment:
Likewise you can get rid of the `if` here with a function like

`_FormatDir(entry, home_dir=None)`